frontend/appflowy_flutter/lib/plugins/database/board/presentation/widgets/board_focus_scope.dart

```diff
@@ -0,0 +1,455 @@
+
+      if (firstGroupIndex == -1 || lastGroupIndex == -1) {
+        _focusLastCard();
+        notifyListeners();
+        return;
+      }
+
+      if (firstGroupIndex < lastGroupIndex) {
+        isExpand = false;
+      } else if (firstGroupIndex > lastGroupIndex) {
+        isExpand = true;
+      } else {
+        final groupItems =
+            boardController.groupDatas.elementAt(firstGroupIndex).items;
+        final firstCardIndex =
+            groupItems.indexWhere((item) => item.id == firstFocusedCard.rowId);
+        final lastCardIndex =
+            groupItems.indexWhere((item) => item.id == lastFocusedCard.rowId);
+
+        if (firstCardIndex == -1 || lastCardIndex == -1) {
+          _focusLastCard();
+          notifyListeners();
+          return;
+        }
+
+        isExpand = firstCardIndex > lastCardIndex;
+      }
+    }
+
+    if (isExpand) {
+      final groupController =
+          boardController.controller(lastFocusedCard.groupId);
+
+      if (groupController == null) {
+        _focusLastCard();
+        notifyListeners();
+        return;
+      }
+
+      final iterable = groupController.items.reversed
+          .skipWhile((item) => item.id != lastFocusedCard.rowId);
+
+      if (iterable.length == 1) {
+        // focus on the last card in the previous group
+        final group = boardController.groupDatas.reversed
+            .skipWhile((item) => item.id != lastFocusedCard.groupId)
+            .skip(1)
+            .firstWhereOrNull((groupData) => groupData.items.isNotEmpty);
+        if (group != null) {
+          _focusedCards.add(
+            GroupedRowId(
+              rowId: group.items.last.id,
+              groupId: group.id,
+            ),
+          );
+        }
+      } else {
+        _focusedCards.add(
+          GroupedRowId(
+            rowId: iterable.elementAt(1).id,
+            groupId: lastFocusedCard.groupId,
+          ),
+        );
+      }
+    } else {
+      _focusedCards.removeLast();
+    }
+
+    notifyListeners();
+  }
+
+  void clear() {
+    _deepCopy();
+    _focusedCards.clear();
+    notifyListeners();
+  }
+
+  void _focusFirstCard() {
+    _focusedCards.clear();
+    final firstGroup = boardController.groupDatas
+        .firstWhereOrNull((group) => group.items.isNotEmpty);
+    final firstCard = firstGroup?.items.firstOrNull;
+    if (firstCard != null) {
+      _focusedCards
+          .add(GroupedRowId(rowId: firstCard.id, groupId: firstGroup!.id));
+    }
+  }
+
+  void _focusLastCard() {
+    _focusedCards.clear();
+    final lastGroup = boardController.groupDatas
+        .lastWhereOrNull((group) => group.items.isNotEmpty);
+    final lastCard = lastGroup?.items.lastOrNull;
+    if (lastCard != null) {
+      _focusedCards
+          .add(GroupedRowId(rowId: lastCard.id, groupId: lastGroup!.id));
+    }
+  }
+
+  void _deepCopy() {
+    _focusedCards = [..._focusedCards];
+  }
+}
+
+class ConditionalListenableBuilder<T> extends StatefulWidget {
```

Comment:
Let's move this to `appflowy_flutter/lib/shared/conditional_listenable_builder.dart`

Reply:
Done!